psycopg2.ProgrammingError: can't adapt type 'dict'

when running job synch from postgres to postgres, error because one column is json.
When ignoring the column in the select statement, the job works 

<details><summary>Job Logs</summary>
 python -m src.main --jobs p2p-order_quotes
2024-12-25 16:50:43 [INFO] dune-sync:job.run:80 - Fetching data for job: p2p-order_quotes
2024-12-25 16:50:43 [INFO] dune-sync:job.run:84 - Saving data for job: p2p-order_quotes
2024-12-25 16:50:44 [ERROR] dune-sync:main.main:44 - Error in job execution: (psycopg2.ProgrammingError) can't adapt type 'dict'
[SQL: INSERT INTO aurelie.backend_data__order_quotes (order_uid, gas_amount, gas_price, sell_token_price, sell_amount, buy_amount, solver, verified, metadata) VALUES (%(order_uid__0)s, %(gas_amount__0)s, %(gas_price__0)s, %(sell_token_price__0)s, %(sell_am ... 186818 characters truncated ... s, %(sell_amount__999)s, %(buy_amount__999)s, %(solver__999)s, %(verified__999)s, %(metadata__999)s)]
[parameters: {'gas_amount__0': 191384.0, 'verified__0': False, 'gas_price__0': 10565971926.875, 'sell_amount__0': 2.3621894801637565e+18, 'order_uid__0': '0x1f5e3131b2c535b3dcb7fb0d24eadae7a19623cc9b92711f75bc0c52ee7fab4fba3eff3dd2b4442b6c141c5008260dc18824f2f16711aed6', 'metadata__0': {}, 'sell_token_price__0': 1.1821166340042355, 'solver__0': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__0': 2.791423152045949e+18, 'gas_amount__1': 301339.0, 'verified__1': False, 'gas_price__1': 10565971926.875, 'sell_amount__1': 7.731425305141209e+18, 'order_uid__1': '0x0f04cf8288d2e2b92fb476daaaf3286810febfa5b1ba3ed51f4f11d9ee231a7cba3eff3dd2b4442b6c141c5008260dc18824f2f16711aee2', 'metadata__1': {}, 'sell_token_price__1': 1.0499386906111836, 'solver__1': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__1': 6.871443616471224e+18, 'gas_amount__2': 267863.0, 'verified__2': False, 'gas_price__2': 10318980861.0, 'sell_amount__2': 5.003001801080648e+21, 'order_uid__2': '0x65bced6903168343c172ca433a76c10665c3563881dc59eda50d78671f1881d88dd0d83783cdda5b2900310e6834a650ddaa21246711af08', 'metadata__2': {}, 'sell_token_price__2': 0.00038372531297756397, 'solver__2': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__2': 4.99806504161372e+21, 'gas_amount__3': 165277.0, 'verified__3': False, 'gas_price__3': 10614307640.999235, 'sell_amount__3': 4.214839471567676e+16, 'order_uid__3': '0xfceffe5cfb1697572325b1f1e5cdee9b3ca5bf5ce8be2deab0c858dffc1015d3d7cb8cc1b56356bb7b78d02e785ead28e21586606711a2a8', 'metadata__3': {}, 'sell_token_price__3': 1.0, 'solver__3': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__3': 8.091034541006356e+19, 'gas_amount__4': 336859.0, 'verified__4': False, 'gas_price__4': 9743336458.212614, 'sell_amount__4': 4.195422619098704e+18, 'order_uid__4': '0xf87e0967bb0b257b6b74f093f1c924dbf33d17a2bf01778d3800d1725557eb7a67e0b8064241f0d784a42782aadd3b480d8cb4296711a915', 'metadata__4': {}, 'sell_token_price__4': 1.0, 'solver__4': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__4': 1593249343041.0, 'gas_amount__5': 200963.0, 'verified__5': False, 'gas_price__5': 10939791284.688337, 'sell_amount__5': 2.1881791280073925e+21, 'order_uid__5': '0x46e3bda804a6b5b1be28c1851ba3b84cab83aee395e2dc1eea3530e73b0486f3f0d420a73e8eb488ab39a7cbe82c2d0c7da288cb6711a37e' ... 8900 parameters truncated ... 'order_uid__994': '0xd2d8543f8ca2e91912816ad020f75861bb175745b438e277c5ba6f060916d26b0d9655ae8627b4d5a7b356e24bcf69ca34cbf0446700e51a', 'metadata__994': {}, 'sell_token_price__994': 1.0, 'solver__994': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__994': 7.169029157319265e+16, 'gas_amount__995': 166387.0, 'verified__995': False, 'gas_price__995': 2750600025.6971774, 'sell_amount__995': 8.925518646563182e+17, 'order_uid__995': '0x456596bfa1c1f794914ae5f6f5ff94283b409b3ed61278e3a1a67697a005b98d9941fd7db2003308e7ee17b04400012278f12ac66700df40', 'metadata__995': {}, 'sell_token_price__995': 0.004613233205417498, 'solver__995': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__995': 4111924656241279.0, 'gas_amount__996': 161824.0, 'verified__996': False, 'gas_price__996': 2923964959.425067, 'sell_amount__996': 5.270181761129647e+16, 'order_uid__996': '0x3906fcf2d7ec78aef1e686f0f26cadddcb792c7908360a048be5e225c46274ff079c868f97aed8e0d03f11e1529c3b056ff21cea6700df40', 'metadata__996': {}, 'sell_token_price__996': 1.0, 'solver__996': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__996': 1.2868754898369456e+21, 'gas_amount__997': 241207.0, 'verified__997': False, 'gas_price__997': 2824910436.0542912, 'sell_amount__997': 1e+20, 'order_uid__997': '0x13fd3cde815c786008c491aabf59cf4dada9b2f731f540692e3603acc29bc03de8f154a43bebdf9ff6dc1dc0b09da94290cce42e6700df40', 'metadata__997': {}, 'sell_token_price__997': 0.000456823800218567, 'solver__997': '0x755bae1cd46c9c27a3230aef0ce923bda13d29f7', 'buy_amount__997': 9.927524385771566e+19, 'gas_amount__998': 371286.0, 'verified__998': False, 'gas_price__998': 2889530193.3272314, 'sell_amount__998': 2.246637315136506e+19, 'order_uid__998': '0x0b20cefb3c5a44a8d7df8af9e7c775d7985ee1c3e232e7d8bffaf1cdfaf13d82097ccdce5d7400861b7b014118427cf550766a476700e573', 'metadata__998': {}, 'sell_token_price__998': 0.009118792087225257, 'solver__998': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__998': 495070639.0, 'gas_amount__999': 201413.0, 'verified__999': False, 'gas_price__999': 3031304083.0, 'sell_amount__999': 5.71396e+18, 'order_uid__999': '0x7737d227c11bbb7a9bb4d3c07a175b4187419e2f69da51e33fe24fbaf80967e10646aab136bd80ac1ec82beb9719753bf30ca5256700e5fa', 'metadata__999': {}, 'sell_token_price__999': 0.001418444453733439, 'solver__999': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__999': 8077286506597146.0}]
(Background on this error at: https://sqlalche.me/e/20/f405)
Traceback (most recent call last):
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2118, in _exec_insertmany_context
    dialect.do_execute(
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
psycopg2.ProgrammingError: can't adapt type 'dict'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/workspaces/dune-sync/src/main.py", line 58, in <module>
    asyncio.run(main(jobs_to_run))
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/workspaces/dune-sync/src/main.py", line 45, in main
    raise e
  File "/workspaces/dune-sync/src/main.py", line 42, in main
    await completed_task
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 631, in _wait_for_one
    return f.result()  # May raise f.exception().
           ^^^^^^^^^^
  File "/workspaces/dune-sync/src/metrics.py", line 64, in wrapper
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/dune-sync/src/job.py", line 86, in run
    affected_rows = self.destination.save(df)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/dune-sync/src/destinations/postgres.py", line 158, in save
    affected_rows = self.append(data)
                    ^^^^^^^^^^^^^^^^^
  File "/workspaces/dune-sync/src/destinations/postgres.py", line 190, in append
    df.to_sql(
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/util/_decorators.py", line 333, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/core/generic.py", line 3087, in to_sql
    return sql.to_sql(
           ^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 842, in to_sql
    return pandas_sql.to_sql(
           ^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 2018, in to_sql
    total_inserted = sql_engine.insert_records(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 1567, in insert_records
    raise err
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 1558, in insert_records
    return table.insert(chunksize=chunksize, method=method)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 1119, in insert
    num_inserted = exec_insert(conn, keys, chunk_iter)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/pandas/io/sql.py", line 1010, in _execute_insert
    result = conn.execute(self.table.insert(), data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_insertmany_context(dialect, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2126, in _exec_insertmany_context
    self._handle_dbapi_exception(
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2355, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2118, in _exec_insertmany_context
    dialect.do_execute(
  File "/home/vscode/.cache/pypoetry/virtualenvs/dune-sync-7Q6K8-36-py3.12/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'dict'
[SQL: INSERT INTO aurelie.backend_data__order_quotes (order_uid, gas_amount, gas_price, sell_token_price, sell_amount, buy_amount, solver, verified, metadata) VALUES (%(order_uid__0)s, %(gas_amount__0)s, %(gas_price__0)s, %(sell_token_price__0)s, %(sell_am ... 186818 characters truncated ... s, %(sell_amount__999)s, %(buy_amount__999)s, %(solver__999)s, %(verified__999)s, %(metadata__999)s)]
[parameters: {'gas_amount__0': 191384.0, 'verified__0': False, 'gas_price__0': 10565971926.875, 'sell_amount__0': 2.3621894801637565e+18, 'order_uid__0': '0x1f5e3131b2c535b3dcb7fb0d24eadae7a19623cc9b92711f75bc0c52ee7fab4fba3eff3dd2b4442b6c141c5008260dc18824f2f16711aed6', 'metadata__0': {}, 'sell_token_price__0': 1.1821166340042355, 'solver__0': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__0': 2.791423152045949e+18, 'gas_amount__1': 301339.0, 'verified__1': False, 'gas_price__1': 10565971926.875, 'sell_amount__1': 7.731425305141209e+18, 'order_uid__1': '0x0f04cf8288d2e2b92fb476daaaf3286810febfa5b1ba3ed51f4f11d9ee231a7cba3eff3dd2b4442b6c141c5008260dc18824f2f16711aee2', 'metadata__1': {}, 'sell_token_price__1': 1.0499386906111836, 'solver__1': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__1': 6.871443616471224e+18, 'gas_amount__2': 267863.0, 'verified__2': False, 'gas_price__2': 10318980861.0, 'sell_amount__2': 5.003001801080648e+21, 'order_uid__2': '0x65bced6903168343c172ca433a76c10665c3563881dc59eda50d78671f1881d88dd0d83783cdda5b2900310e6834a650ddaa21246711af08', 'metadata__2': {}, 'sell_token_price__2': 0.00038372531297756397, 'solver__2': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__2': 4.99806504161372e+21, 'gas_amount__3': 165277.0, 'verified__3': False, 'gas_price__3': 10614307640.999235, 'sell_amount__3': 4.214839471567676e+16, 'order_uid__3': '0xfceffe5cfb1697572325b1f1e5cdee9b3ca5bf5ce8be2deab0c858dffc1015d3d7cb8cc1b56356bb7b78d02e785ead28e21586606711a2a8', 'metadata__3': {}, 'sell_token_price__3': 1.0, 'solver__3': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__3': 8.091034541006356e+19, 'gas_amount__4': 336859.0, 'verified__4': False, 'gas_price__4': 9743336458.212614, 'sell_amount__4': 4.195422619098704e+18, 'order_uid__4': '0xf87e0967bb0b257b6b74f093f1c924dbf33d17a2bf01778d3800d1725557eb7a67e0b8064241f0d784a42782aadd3b480d8cb4296711a915', 'metadata__4': {}, 'sell_token_price__4': 1.0, 'solver__4': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__4': 1593249343041.0, 'gas_amount__5': 200963.0, 'verified__5': False, 'gas_price__5': 10939791284.688337, 'sell_amount__5': 2.1881791280073925e+21, 'order_uid__5': '0x46e3bda804a6b5b1be28c1851ba3b84cab83aee395e2dc1eea3530e73b0486f3f0d420a73e8eb488ab39a7cbe82c2d0c7da288cb6711a37e' ... 8900 parameters truncated ... 'order_uid__994': '0xd2d8543f8ca2e91912816ad020f75861bb175745b438e277c5ba6f060916d26b0d9655ae8627b4d5a7b356e24bcf69ca34cbf0446700e51a', 'metadata__994': {}, 'sell_token_price__994': 1.0, 'solver__994': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__994': 7.169029157319265e+16, 'gas_amount__995': 166387.0, 'verified__995': False, 'gas_price__995': 2750600025.6971774, 'sell_amount__995': 8.925518646563182e+17, 'order_uid__995': '0x456596bfa1c1f794914ae5f6f5ff94283b409b3ed61278e3a1a67697a005b98d9941fd7db2003308e7ee17b04400012278f12ac66700df40', 'metadata__995': {}, 'sell_token_price__995': 0.004613233205417498, 'solver__995': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__995': 4111924656241279.0, 'gas_amount__996': 161824.0, 'verified__996': False, 'gas_price__996': 2923964959.425067, 'sell_amount__996': 5.270181761129647e+16, 'order_uid__996': '0x3906fcf2d7ec78aef1e686f0f26cadddcb792c7908360a048be5e225c46274ff079c868f97aed8e0d03f11e1529c3b056ff21cea6700df40', 'metadata__996': {}, 'sell_token_price__996': 1.0, 'solver__996': '0x4339889fd9dfca20a423fba011e9dff1c856caeb', 'buy_amount__996': 1.2868754898369456e+21, 'gas_amount__997': 241207.0, 'verified__997': False, 'gas_price__997': 2824910436.0542912, 'sell_amount__997': 1e+20, 'order_uid__997': '0x13fd3cde815c786008c491aabf59cf4dada9b2f731f540692e3603acc29bc03de8f154a43bebdf9ff6dc1dc0b09da94290cce42e6700df40', 'metadata__997': {}, 'sell_token_price__997': 0.000456823800218567, 'solver__997': '0x755bae1cd46c9c27a3230aef0ce923bda13d29f7', 'buy_amount__997': 9.927524385771566e+19, 'gas_amount__998': 371286.0, 'verified__998': False, 'gas_price__998': 2889530193.3272314, 'sell_amount__998': 2.246637315136506e+19, 'order_uid__998': '0x0b20cefb3c5a44a8d7df8af9e7c775d7985ee1c3e232e7d8bffaf1cdfaf13d82097ccdce5d7400861b7b014118427cf550766a476700e573', 'metadata__998': {}, 'sell_token_price__998': 0.009118792087225257, 'solver__998': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__998': 495070639.0, 'gas_amount__999': 201413.0, 'verified__999': False, 'gas_price__999': 3031304083.0, 'sell_amount__999': 5.71396e+18, 'order_uid__999': '0x7737d227c11bbb7a9bb4d3c07a175b4187419e2f69da51e33fe24fbaf80967e10646aab136bd80ac1ec82beb9719753bf30ca5256700e5fa', 'metadata__999': {}, 'sell_token_price__999': 0.001418444453733439, 'solver__999': '0x008300082c3000009e63680088f8c7f4d3ff2e87', 'buy_amount__999': 8077286506597146.0}]
(Background on this error at: https://sqlalche.me/e/20/f405)

</details>

--- a/config.yaml
+++ b/config.yaml
@@ -56,7 +56,8 @@ jobs:
         SELECT 
           1 AS number, 
           '\\x1234'::bytea AS my_bytes,
-          '{"key": "value", "array": [1, 2, 3], "dict": {}}'::json AS my_json
+          '{"key": "value", "array": [1, 2, 3], "dict": {}}'::json AS my_json,
+          '[ [{"x": 1}, {"y": 2}], null, [{"z": 3}] ]'::json AS list_dict
 
     destination:
       ref: PG

--- a/config.yaml
+++ b/config.yaml
@@ -52,7 +52,12 @@ jobs:
   - name: p2p-test
     source:
       ref: PG
-      query_string: "SELECT 1 as number, '\\x1234'::bytea as my_bytes;"
+      query_string: |
+        SELECT 
+          1 AS number, 
+          '\\x1234'::bytea AS my_bytes,
+          '{"key": "value", "array": [1, 2, 3], "dict": {}}'::json AS my_json
+
     destination:
       ref: PG
       table_name: moo.p2p-test

--- a/src/sources/postgres.py
+++ b/src/sources/postgres.py
@@ -15,11 +15,20 @@ from src.logger import log
 
 
 def _convert_dict_to_json(df: DataFrame) -> DataFrame:
-    """Convert dictionary columns to JSON strings."""
+    if df.empty:
+        return df
+
     df = df.copy()
     for column in df.columns:
-        if isinstance(df[column].iloc[0], dict):
-            df[column] = df[column].apply(json.dumps)
+        # Get first non-null value to check type
+        non_null_values = df[column][df[column].notna()]
+        if len(non_null_values) > 0:
+            first_value = non_null_values.iloc[0]
+            if isinstance(first_value, dict | list):
+                df[column] = df[column].apply(
+                    lambda x: json.dumps(x) if x is not None else None
+                )
+
     return df
 
 

--- a/src/sources/postgres.py
+++ b/src/sources/postgres.py
@@ -1,6 +1,7 @@
 """Source logic for PostgreSQL."""
 
 import asyncio
+import json
 from pathlib import Path
 
 import pandas as pd
@@ -11,6 +12,15 @@ from sqlalchemy.exc import SQLAlchemyError
 
 from src.interfaces import Source, TypedDataFrame
 from src.logger import log
+
+
+def _convert_dict_to_json(df: DataFrame) -> DataFrame:
+    """Convert dictionary columns to JSON strings."""
+    df = df.copy()
+    for column in df.columns:
+        if isinstance(df[column].iloc[0], dict):
+            df[column] = df[column].apply(json.dumps)
+    return df
 
 
 def _convert_bytea_to_hex(df: DataFrame) -> DataFrame:
@@ -121,8 +131,11 @@ class PostgresSource(Source[TypedDataFrame]):
         df = await loop.run_in_executor(
             None, lambda: pd.read_sql_query(self.query_string, con=self.engine)
         )
+
+        df = _convert_dict_to_json(df)
+        df = _convert_bytea_to_hex(df)
         # TODO include types.
-        return TypedDataFrame(dataframe=_convert_bytea_to_hex(df), types={})
+        return TypedDataFrame(dataframe=df, types={})
 
     def is_empty(self, data: TypedDataFrame) -> bool:
         """Check if the provided DataFrame is empty.

--- a/tests/unit/sources_test.py
+++ b/tests/unit/sources_test.py
@@ -11,7 +11,11 @@ from sqlalchemy.dialects.postgresql import BYTEA
 from src.config import RuntimeConfig
 from src.interfaces import TypedDataFrame
 from src.sources.dune import _reformat_varbinary_columns, dune_result_to_df
-from src.sources.postgres import PostgresSource, _convert_bytea_to_hex
+from src.sources.postgres import (
+    PostgresSource,
+    _convert_bytea_to_hex,
+    _convert_dict_to_json,
+)
 from tests import config_root, fixtures_root
 
 INVALID_CONFIG_MESSAGE = (
@@ -76,6 +80,35 @@ class TestSourceUtils(unittest.TestCase):
         df = pd.DataFrame([])
         result = _convert_bytea_to_hex(df)
         pd.testing.assert_frame_equal(pd.DataFrame([]), result)
+
+    def test_convert_dict_to_json(self):
+        # Test data with dictionary and normal columns
+        data = {
+            "dict_col": [{"key": "value"}, {"nested": {"a": 1}}, None],
+            "normal_col": [1, 2, 3],
+            "list_dict": [[{"x": 1}, {"y": 2}], None, [{"z": 3}]],
+        }
+        df = pd.DataFrame(data)
+
+        result = _convert_dict_to_json(df)
+
+        # Assert dictionary column was converted to JSON strings
+        assert result["dict_col"].tolist() == [
+            '{"key": "value"}',
+            '{"nested": {"a": 1}}',
+            None,
+        ]
+
+        # Assert list of dictionaries was converted
+        assert result["list_dict"].tolist() == ['[{"x": 1}, {"y": 2}]', None, '[{"z": 3}]']
+
+        # Assert normal column remains unchanged
+        assert result["normal_col"].tolist() == [1, 2, 3]
+
+        # Test empty DataFrame
+        df_empty = pd.DataFrame([])
+        result_empty = _convert_dict_to_json(df_empty)
+        pd.testing.assert_frame_equal(pd.DataFrame([]), result_empty)
 
 
 class TestPostgresSource(unittest.TestCase):

--- a/tests/unit/sources_test.py
+++ b/tests/unit/sources_test.py
@@ -100,7 +100,11 @@ class TestSourceUtils(unittest.TestCase):
         ]
 
         # Assert list of dictionaries was converted
-        assert result["list_dict"].tolist() == ['[{"x": 1}, {"y": 2}]', None, '[{"z": 3}]']
+        assert result["list_dict"].tolist() == [
+            '[{"x": 1}, {"y": 2}]',
+            None,
+            '[{"z": 3}]',
+        ]
 
         # Assert normal column remains unchanged
         assert result["normal_col"].tolist() == [1, 2, 3]

